Overhaul UI

boerdereinar opened the pull request:
# Overhaul UI

## Description
This pr fully overhauls the UI as per #185.

## New Features
- Complete overhaul of UI
  - Adwaita colors (customisation is still possible)
  - Favorite/close button only visible on hover (configurable)
  - Pano window floats
    - Can be disabled
    - Each margin can be configured
  - Configurable item width and height
- File previews
  - Text files
  - Images
  - Files that have a generated thumbnail i.e. pdfs/videos
  - File paths containing `/home/user/*` are shortened to `~/*`
  - Multiple copied files are listed under their common directory to shorten the paths of the copied files. Usually this will result in only the file names being shown
  - Color item foreground is computed using luminance
- Header Styles
  - Headers are hidden by default
  - Visible
  - Compact
- Compact Mode
  - Height of item is halved
  - Layout of items is restructured
  - File previews are hidden
- Open at pointer
  - Pano can open at the current pointer position
  - Height of the window is configurable
- Middle click to delete items
- Paste file content
- Items are no longer selected after deleting an item (if no items were selected before)
- Any single grapheme cluster is now displayed as an emoji

## Misc
- `yarn run clean:ts` can now fail in `build:ts` to make building easier while pano is still running

## Dependencies
- Added `color-string` for easy color parsing and conversion to rgb
- Added `graphemer` for checking if the copied string contains a single grapheme cluster for the emoji item. This fixes ZWJ sequences not being detected as an emoji
- Removed `validate-color`
- Removed `hex-color-converter`

## Screenshots
### New UI
![image](https://github.com/oae/gnome-shell-pano/assets/54003172/6a130695-7854-4eaf-9dd8-d94c91352236)

### Compact Mode
![image](https://github.com/oae/gnome-shell-pano/assets/54003172/e3c75b5c-9222-4aa8-9b46-495c2db832d9)

### Window Position: Pointer
![image](https://github.com/oae/gnome-shell-pano/assets/54003172/33e76271-66f6-4967-a83d-bbefdf3b4430)

### Incognito Mode
![image](https://github.com/oae/gnome-shell-pano/assets/54003172/7f177fdc-c801-4df0-abe8-fa9790237ac7)

### Header Style: Compact
![image](https://github.com/oae/gnome-shell-pano/assets/54003172/d5985070-7755-4226-93b1-5082c20a2935)

## Fixes
Fixes #124 
Fixes #150 
Fixes #176 
Fixes #185 
Fixes #272 
Fixes #300 

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] My commits follow the commit standards of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings

boerdereinar committed:
Change to headerless style + add compact mode

boerdereinar committed:
Remove redundant settings

boerdereinar committed:
Shorten /home/user to ~

boerdereinar committed:
add back headers

boerdereinar committed:
add close item with middle click

boerdereinar committed:
Add file preview + various style fixes

boerdereinar committed:
Add file type detection for files without an extension

boerdereinar committed:
Modify file item look

boerdereinar committed:
Change file styling again

boerdereinar committed:
Rename preview settings

boerdereinar committed:
Don't select on item removal

boerdereinar committed:
Replace validate-color calls with color-string + make clean:ts failable in build:ts + style changes

clean:ts is now able to fail so that I can build Pano while it is still running

boerdereinar committed:
Add color-string to rollup

boerdereinar committed:
Fix prettier problems

boerdereinar committed:
Re-add removed translations

boerdereinar committed:
Make item width and height configurable

boerdereinar committed:
Change text styling

boerdereinar committed:
Only parse rgb, because color-string doesn't do conversions

boerdereinar committed:
Fix compact mode link style

Totto16 commented:
Thank you for your contribution, I don't know if it makes sense to review this yet, since this and #278 clash and might create merge conflicts. I also am in favor of publishing the already working UI with #278 to EGO first, before during this. And then testing it extensively and by multiple users, before releasing it. 


But overall it really looks good, so I'm going to look at this as soon as the issues and PR's with more priority are resolved. 

And just from glancing over it, there are a few things to do:
- [ ] replace images / videos in the README with the new design (do that at the end, since we might change something)

Totto16 commented on package.json:
```diff
@@ -7,7 +7,7 @@
     "build": "yarn run build:ts && yarn run gen:locale && yarn run build:extension",
     "gen:locale": "yarn run ts-node scripts/generateLocale.ts",
-    "build:ts": "yarn run clean:ts && rollup -c --failAfterWarnings && sed -i '/setTimeout/d' dist/thirdparty/prismjs.js && sed -i 's/var\\ /let\\ /g' dist/extension.js dist/prefs.js",
+    "build:ts": "yarn run clean:ts; rollup -c --failAfterWarnings && sed -i '/setTimeout/d' dist/thirdparty/prismjs.js && sed -i 's/var\\ /let\\ /g' dist/extension.js dist/prefs.js",
```
This is just here to block this from being merged for the moment, please leave this unresolved

Totto16 replied:
Why was this changed from  `&&` to `;` did `yarn run clean:ts` fail locally? since it would be better, if the rest only gets executed, if that also works.

boerdereinar committed:
Add floating mode + fix incognito icon + modify colors to better support incognito mode

boerdereinar committed:
Use BinLayout for pano item

boerdereinar committed:
Add option to open Pano at the pointer position

boerdereinar committed:
Fix action button coloring to follow body/header color

boerdereinar committed:
Add compact header style

boerdereinar committed:
Fix item spacing evenly

boerdereinar committed:
Reduce margins of compact text item to fit more lines

boerdereinar committed:
Only disable pointer location for indicator

boerdereinar committed:
Fix link button background

boerdereinar committed:
Fix jumpy animations when filtering/deleting items + fix item spacing again

boerdereinar committed:
Move file icons to header icon if headers are enabled

boerdereinar committed:
Fix metadata height for link item with header

boerdereinar committed:
Fix window height not updating + update resource files

boerdereinar committed:
Add favorite indicator + modify file icon placement and link preview style in compact mode

boerdereinar committed:
Fix shortcuts

boerdereinar committed:
Add paste file content functionality to text file items

Can be triggered by:
1. Pressing the action button
2. Ctrl + click on the item
3. Ctrl + enter on the item

boerdereinar committed:
Store any single grapheme as an emoji item

This adds support for ZWJ sequences

boerdereinar committed:
Fix transparent header color

Note: this causes a windows xp like visual bug in the link item when running Pano in a nested shell

boerdereinar committed:
Decrease height of files preview if it contains less files than the height

boerdereinar committed:
Fix emoji display for wide emojis like 👩🏽‍👩🏿‍👧🏻‍👧🏽

boerdereinar committed:
Fix header date updater not being destroyed

boerdereinar committed:
Fix warnings that happen when toggling the show-controls-on-hover setting

boerdereinar committed:
Add border around window

boerdereinar committed:
Fix item shadow + color item color + remove on mmb setting not being respected

- Item color didn't support this format: rgb(255 0 0)

boerdereinar committed:
Fix pixel inconsistencies

JE-Amouel commented:
Hi,
Is this PR still in the scope?

PauloVBettio commented:
What is missing to merge this pull request? How could I help to speed this up?

JE-Amouel commented:
i just try this branch, and is f*cking awesome :100:

JE-Amouel commented:
@PauloVBettio 
>So, it’s currently broken? Bummer.

If you're talking about the artifacts on the screenshot, it's Just because of my screenshot tool which, I think, only works well under wayland, 
otherwise I spent a little afternoon 😎 testing the different features, and I opted for a centralized interface like on the screenshot ,

 >I could also submit a new pull request with the design, but would it actually get merged? If not, I think it might be better to create a new fork or start a new project with a different vision.

If your code is hot, I think it's worth a try before launching into a new branch, I would also like to take advantage of the progress of the main branch which has Added highlighting features for code. And also without mentioning all the refactoring necessary to adapt it to the new version of gnome, but frankly if you are determined enough to launch a new branch or even a new extension with these features, it's really worth it.✨👌

>No problem with Pano tho, it is great! I understand if they don't want to change the design.

😵‍💫

boerdereinar commented:
Afaik everything works, that screenshot just has the background set to transparent and the headers enabled. I just haven't updated it since the current branch still works fine.

PauloVBettio commented:
Sorry, I thought it was broken because of the artifacts and headers, my bad, how embarrassing 😅
Even so, is there anything I can do to speed up the merge process? I'm willing to work on this because I really like the work that was done in this PR.

JE-Amouel commented:
@PauloVBettio 
>Sorry, I thought it was broken because of the artifacts and headers, my bad, how embarrassing 😅
:laughing: , sorry, my bad...

I send a better screenshot

![Screenshot from 2025-01-24 18-35-24](https://github.com/user-attachments/assets/b3212514-3b1e-4c01-929e-75ef711d0ab1)

I think there is still some work to be done, because there are sometimes images that are not displayed.
but still great.
>is there anything I can do to speed up the merge process? I'm willing to work on this because I really like the work that was done in this PR.

since there no any activities concerning this PR, it difficult to know :confused:

oae commented:
@boerdereinar could you please resolve the conflicts so I can merge them?

Totto16 commented on resources/icons/hicolor/scalable/actions/blend-tool-symbolic.svg:
General question regarding new icons, is there no adwaita-icon from https://github.com/GNOME/adwaita-icon-theme that is similar and could be used instead?

Totto16 reviewed (approved):
The code looks good from my side. Thanks for this amazing UI overhaul ❤️

Totto16 commented:
> @boerdereinar could you please resolve the conflicts so I can merge them?

Since he ticked the box `Maintainers are allowed to edit this pull request` I'll just resolve the merge conflicts, one thing that needs to be done (but can be after this PR is merged) is the updating of the screenshots in the README

Totto16 commented:
I rebased the branch on master, there were many merge conflicts, that were not that easy to resolve, so I made a backup of the changes before the rebase on https://github.com/oae/gnome-shell-pano/tree/gnome-theming-v2_before_rebase

I think I did it correctly and everything is as expected, but I might have made a mistake. Feel free to modify this or force push (`--force-with-lease` to be sure) your own rebase or edits 😄

Totto16 committed:
fix: use orientation compatibility functions instead of setting vertical (which is being deprecated)

oae commented on resources/schemas/org.gnome.shell.extensions.pano.gschema.xml:
There are some default value changes in gschema. If people have already customized it before, these changes will not take effect on their system, right? Maybe we should change the keys or the schema ID. What do you think, @Totto16?

Totto16 replied:
Yes that would be a good idea, as someone, who didn't change most of the settings in didn't notice it.

Changing the schema ID seems to be a good idea, so that incorporates some version or something of that sort 🤔   

My first thought would be `id="org.gnome.shell.extensions.pano.v2"` or something similar, what do you think?

oae replied:
Also, since accent colors are a thing now, maybe we can adapt them for these two customization options
![image](https://github.com/user-attachments/assets/0865db26-3696-477b-aa6d-7aa191632610)

Totto16 replied:
That would also be great, but I think it is out of the scope of this PR, as accent colors are only introduced as of gnome 47, if I remember correctly. And we need to listen for changes correctly, and interpret colors correctly, since as far as I know, some distros (and maybe also stock gnome) refers to accent colors by name and not by rgb or hex color 🤔

oae replied:
After the schema change, we can merge this.

Totto16 reviewed (changes requested):
Please incorporate the schema id change as discussed in https://github.com/oae/gnome-shell-pano/pull/280#discussion_r1966555192

radiolamp commented:
Great job, would it be possible to switch the color scheme in gnome depending on the light/dark theme?

Totto16 commented:
> Great job, would it be possible to switch the color scheme in gnome depending on the light/dark theme?

That is not that easy, as there are many customizations, that can't just be applied to the dark / light heme. But it's generally a good idea.

mbeps commented:
So I am guessing this is almost finished?

JE-Amouel commented:
Hi, i would really like to set two things, so for me pano will be totally completed, first is the search entry if is not possible to dynamically display it, is it possible to add option to set is position to bottom and also change the border color, so I can make it presence less visible when I don't need it, seconde is the border radius of the content preview, thanks

ClairDePrince commented:
Hello! Any update on this?

tanaybhomia commented:
Any reason why this is not being merged or can someone tell me how can I get the new updated UI ? Which branch because when I tried the @boerdereinar branch I had errors with the chalk dependency

AshtakaOOf commented:
> Any reason why this is not being merged

As you can see there have been schema changes, and the author of the PR isn't active to fix them.

mbeps commented:
> > Any reason why this is not being merged
> 
> As you can see there have been schema changes, and the author of the PR isn't active to fix them.

This is bummer because the author did a significant amount of work on this.

JE-Amouel commented:
author of this PR just release it own version of pano 
[copyus](https://github.com/boerdereinar/copyous)

ClairDePrince commented:
> author of this PR just release it own version of pano [copyus](https://github.com/boerdereinar/copyous)

Thanks!
